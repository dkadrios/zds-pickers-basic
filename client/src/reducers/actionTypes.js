export default {
  ADD_ENTRY: 'ADD_ENTRY',
  CANCEL_ENTRY_EDIT: 'CANCEL_ENTRY_EDIT',
  CHANGE_CONTROL_TYPE: 'CHANGE_CONTROL_TYPE',
  CHANGE_ENTRY_CHANNEL: 'CHANGE_ENTRY_CHANNEL',
  CHANGE_ENTRY_STATUS: 'CHANGE_ENTRY_STATUS',
  CHANGE_ENTRY_VALUE: 'CHANGE_ENTRY_VALUE',
  CHANGE_GROUP_CHANNEL: 'CHANGE_GROUP_CHANNEL',
  CHANGE_GROUP_VALUE: 'CHANGE_GROUP_VALUE',
  CHANGE_INPUT_CONTROL_CHANNEL: 'CHANGE_INPUT_CONTROL_CHANNEL',
  CHANGE_INPUT_CONTROL_LATCHING: 'CHANGE_INPUT_CONTROL_LATCHING',
  CHANGE_INPUT_CONTROL_POLARITY: 'CHANGE_INPUT_CONTROL_POLARITY',
  CHANGE_INPUT_CONTROL_VALUE: 'CHANGE_INPUT_CONTROL_VALUE',
  CHANGE_SELECTED_GROUP: 'CHANGE_SELECTED_GROUP',
  CHECKED_REGISTRATION: 'CHECKED_REGISTRATION',
  CLEAR_ALL_SELECTIONS: 'CLEAR_ALL_SELECTIONS',
  CONFIRM_FACTORY_RESET: 'CONFIRM_FACTORY_RESET',
  DELETE_MAPPING: 'DELETE_MAPPING',
  DEVICE_REGISTERED: 'DEVICE_REGISTERED',
  DISMISS_ERROR: 'DISMISS_ERROR',
  EDIT_ENTRY: 'EDIT_ENTRY',
  ENTRY_META_CLICKED: 'ENTRY_META_CLICKED',
  EXPORT_SETTINGS_PACKET: 'EXPORT_SETTINGS_PACKET',
  EXPORT_SETTINGS: 'EXPORT_SETTINGS',
  FACTORY_RESET: 'FACTORY_RESET',
  GET_SYSEX_CONTROLS: 'GET_SYSEX_CONTROLS',
  GET_SYSEX_GROUPS: 'GET_SYSEX_GROUPS',
  GET_SYSEX_MODEL: 'GET_SYSEX_MODEL',
  GET_SYSEX_VERSION: 'GET_SYSEX_VERSION',
  HARDWARE_TEST_FACTORY_RESET: 'HARDWARE_TEST_FACTORY_RESET',
  HARDWARE_TEST_NEXT: 'HARDWARE_TEST_NEXT',
  HARDWARE_TEST_SEND_ENTRIES: 'HARDWARE_TEST_SEND_ENTRIES',
  HARDWARE_TEST_SEND_MESSAGE: 'HARDWARE_TEST_SEND_MESSAGE',
  HARDWARE_TEST_SEND_MESSAGES: 'HARDWARE_TEST_SEND_MESSAGES',
  HIDE_HARDWARE_TEST_DIALOG: 'HIDE_HARDWARE_TEST_DIALOG',
  HIDE_REGISTRATION_DLG: 'HIDE_REGISTRATION_DLG',
  HIDE_REGISTRATION_NAG: 'HIDE_REGISTRATION_NAG',
  IMPORT_MAPPING: 'IMPORT_MAPPING',
  IMPORT_SETTINGS: 'IMPORT_SETTINGS',
  INVALID_SETTINGS_FILE_ACK: 'INVALID_SETTINGS_FILE_ACK',
  INVALID_SETTINGS_FILE: 'INVALID_SETTINGS_FILE',
  MAPPING_CHANGED: 'MAPPING_CHANGED',
  MIDI_IDLE: 'MIDI_IDLE',
  MIDI_IN_ACTIVITY: 'MIDI_IN_ACTIVITY',
  MIDI_OUT_ACTIVITY: 'MIDI_OUT_ACTIVITY',
  NOT_RESPONDING: 'NOT_RESPONDING',
  RECEIVED_AVAILABILITY: 'RECEIVED_AVAILABILITY',
  RECEIVED_CONTROLS: 'RECEIVED_CONTROLS',
  RECEIVED_ENTRY: 'RECEIVED_ENTRY',
  RECEIVED_GROUP: 'RECEIVED_GROUP',
  RECEIVED_GROUPS: 'RECEIVED_GROUPS',
  RECEIVED_MODEL: 'RECEIVED_MODEL',
  RECEIVED_STATE: 'RECEIVED_STATE',
  RECEIVED_VERSION: 'RECEIVED_VERSION',
  REMOVE_ENTRY: 'REMOVE_ENTRY',
  REPORT_ERROR: 'REPORT_ERROR',
  RESTART: 'RESTART',
  SAVE_ENTRY_EDIT: 'SAVE_ENTRY_EDIT',
  SEARCHED_FOR_SHIFTER: 'SEARCHED_FOR_SHIFTER',
  SELECT_ALL_ENTRIES: 'SELECT_ALL_ENTRIES',
  SELECT_SHIFT_ENTRY: 'SELECT_SHIFT_ENTRY',
  SET_FLAGS: 'SET_FLAGS',
  SHIFTER_FOUND: 'SHIFTER_FOUND',
  SHIFTER_MISSING: 'SHIFTER_MISSING',
  SHOW_EXPORT_DIALOG: 'SHOW_EXPORT_DIALOG',
  SHOW_HARDWARE_TEST_DIALOG: 'SHOW_HARDWARE_TEST_DIALOG',
  SHOW_IMPORT_DIALOG: 'SHOW_IMPORT_DIALOG',
  SHOW_MAPPINGS_DLG: 'SHOW_MAPPINGS_DLG',
  SHOW_REGISTRATION_DLG: 'SHOW_REGISTRATION_DLG',
  SHOW_USER_MAPPINGS_DIALOG: 'SHOW_USER_MAPPINGS_DIALOG',
  SORT_GROUP: 'SORT_GROUP',
  TEST_INTERFACE_FOUND: 'TEST_INTERFACE_FOUND',
  TEST_INTERFACE_MISSING: 'TEST_INTERFACE_MISSING',
}
