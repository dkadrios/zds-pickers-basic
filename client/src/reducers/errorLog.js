import { createReducer } from './utils'
import actionTypes from './actionTypes'

export const initialState = {
  currentMessage: undefined,
  logVisible: false,
  messages: [],
}

export const actions = {
  reportError: message => ({
    type: actionTypes.ERROR_OCCURRED,
    message,
  }),

  clearError: () => ({ type: actionTypes.ERROR_CLEAR }),

  clearAll: () => ({ type: actionTypes.ERROR_CLEAR_ALL }),

  showLog: logVisible => ({ type: actionTypes.ERROR_SHOW_LOG, logVisible }),
}

const handleError = (state, { message }) => {
  const messages = [...state.messages]
  const newMessage = String(message?.message ? message.message : message)

  if (messages.length && messages[messages.length - 1].text === newMessage) {
    messages[messages.length - 1].count += 1
  } else {
    messages.push({ text: newMessage, count: 1 })
  }

  return {
    ...state,
    currentMessage: newMessage,
    messages: [...messages],
  }
}

const clearError = state => ({ ...state, currentMessage: undefined })
const clearAll = state => ({
  ...state,
  currentMessage: undefined,
  logVisible: false,
  messages: [],
})

const handleShow = (state, { logVisible }) => ({
  ...state,
  logVisible,
})

export default createReducer(initialState, {
  [actionTypes.ERROR_OCCURRED]: handleError,
  [actionTypes.ERROR_CLEAR]: clearError,
  [actionTypes.ERROR_CLEAR_ALL]: clearAll,
  [actionTypes.ERROR_SHOW_LOG]: handleShow,
})
