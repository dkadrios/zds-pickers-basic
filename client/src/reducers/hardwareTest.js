/* eslint-disable no-bitwise, no-console */
import {
  MASK_CHANNEL,
  MASK_STATUS,
  STATUS_CONTROL_CHANGE,
  STATUS_NOTE_OFF,
  STATUS_NOTE_ON,
  combineStatus,
} from 'zds-pickers/dist/midi/statuses'
import rules from '../components/device/hardwareTest/rules'
import echoMessages from '../components/device/hardwareTest/echoMessages'
import { shifterInputId, testInterfaceInputId } from '../midi/devices'
import { SHIFTER_CC_MESSAGE } from '../midi'
import { delay } from '../core/fp/utils'
import { isEqual } from '../core/fp/objects'
import { createReducer } from './utils'
import { actions as shifterActions } from './shifter'
import actionTypes from './actionTypes'
import { compareAll, compareValue } from './shiftEntry'

const initialState = {
  step: 0,
  done: false,
  busy: false,
  performingReset: false,
  creatingRules: false,
  sendingEntries: false,
  sendingMessages: false,
  messagesReceived: [],
  passed: true,
}

export const actions = {
  nextStep: () => ({ type: actionTypes.HARDWARE_TEST_NEXT }),

  performFactoryReset: done => (dispatch) => {
    dispatch({ type: actionTypes.HARDWARE_TEST_FACTORY_RESET, done })
    dispatch(shifterActions.performFactoryReset(true))
  },

  sendEntries: busy => (dispatch) => {
    dispatch({ type: actionTypes.HARDWARE_TEST_SEND_ENTRIES, busy })

    rules.forEach(({ groupId, ...rest }) => {
      dispatch({ type: actionTypes.SAVE_ENTRY_EDIT, groupId, editQueue: { ...rest } })
      delay(50)
    })

    // Turn on both buttons
    dispatch({
      type: actionTypes.HARDWARE_TEST_SEND_MESSAGE,
      message: combineStatus(9, STATUS_CONTROL_CHANGE),
      data: [SHIFTER_CC_MESSAGE, 127],
    })

    delay(50)

    dispatch({
      type: actionTypes.HARDWARE_TEST_SEND_MESSAGE,
      message: combineStatus(9, STATUS_CONTROL_CHANGE),
      data: [SHIFTER_CC_MESSAGE + 1, 127],
    })
  },

  sendMessages: () => (dispatch) => {
    dispatch({ type: actionTypes.HARDWARE_TEST_SEND_MESSAGES })
    rules.forEach(({ input: { channel, status, value } }) => {
      dispatch({
        type: actionTypes.HARDWARE_TEST_SEND_MESSAGE,
        message: combineStatus(channel, status),
        data: [value, 127],
      })
      delay(50)
    })

    echoMessages.forEach(({ channel, status, value }) => {
      dispatch({
        type: actionTypes.HARDWARE_TEST_SEND_MESSAGE,
        message: combineStatus(channel, status),
        data: [value, 102],
      })
      delay(50)
    })
  },
}

const analyzeResults = (messagesReceived) => {
  const messagesByDevice = deviceId => messagesReceived //
    .filter(({ device }) => device === deviceId)
    .map(({ device, value, ...rest }) => ({ ...rest }))
    .map(({ noteNum: value, ...rest }) => ({ value, ...rest }))
  const dinMessages = messagesByDevice(testInterfaceInputId())
  const usbMessages = messagesByDevice(shifterInputId())

  console.log('din received', dinMessages)
  console.log('usb received', usbMessages)

  if (dinMessages.length + usbMessages.length !== (rules.length + echoMessages.length) * 2 + 2) {
    console.error('Not enough messages were received!')
    console.log(dinMessages.length + usbMessages.length, (rules.length + echoMessages.length) * 2 + 2)

    return false
  }

  if (!isEqual(dinMessages, usbMessages)) {
    console.error('DIN and USB results are not the same')
    return false
  }

  const expected = [
    //
    ...rules.map(({ output }) => ({ ...output })),
    ...echoMessages,
    ...rules //
      .filter(({ output: { status } }) => status === STATUS_NOTE_OFF)
      .map(({ input }) => ({ ...input })),
  ]
    .map(({ value, channel, status }) => ({
      value,
      channel,
      status: status === STATUS_NOTE_OFF ? STATUS_NOTE_ON : status,
    }))
    .sort(compareAll)
    .sort(compareValue)

  const actual = dinMessages.sort(compareAll).sort(compareValue)

  if (!isEqual(expected, actual)) {
    console.error('Outputs did not match expectations:')
    console.log('expected - actual')
    console.log(expected, actual)
    return false
  }

  return true
}

const reset = () => ({ ...initialState })

const changeStep = inc => state => ({ ...state, step: state.step + inc })

const handleFactoryReset = (state, { done }) => ({
  ...state,
  busy: true,
  performingReset: true,
  done,
})

const handleSendEntries = (state, { busy }) => ({ ...state, busy, creatingRules: true })

const handleIdle = (state) => {
  const newState = {
    ...state,
    busy: false,
    creatingRules: false,
    performingReset: false,
    sendingMessages: false,
  }

  if (state.sendingMessages) {
    newState.passed = analyzeResults(state.messagesReceived)
  }

  return newState
}

const handleSendMessages = state => ({ ...state, busy: true, sendingMessages: true })

const receivedMidiMessage = (state, { payload }) => {
  if (state.sendingMessages) {
    const { data, device } = payload
    const [message, noteNum, value] = data
    const channel = message & MASK_CHANNEL
    const status = ((message & MASK_STATUS) >> 4) | 8
    return {
      ...state,
      messagesReceived: [
        ...state.messagesReceived,
        {
          device,
          channel,
          status,
          noteNum,
          value,
        },
      ],
    }
  }
  return state
}

export default createReducer(initialState, {
  [actionTypes.SHOW_HARDWARE_TEST_DIALOG]: reset,
  [actionTypes.HARDWARE_TEST_NEXT]: changeStep(1),
  [actionTypes.HARDWARE_TEST_FACTORY_RESET]: handleFactoryReset,
  [actionTypes.HARDWARE_TEST_SEND_ENTRIES]: handleSendEntries,
  [actionTypes.MIDI_IDLE]: handleIdle,
  [actionTypes.HARDWARE_TEST_SEND_MESSAGES]: handleSendMessages,
  'redux-midi/midi/RECEIVE_MIDI_MESSAGE': receivedMidiMessage,
})
