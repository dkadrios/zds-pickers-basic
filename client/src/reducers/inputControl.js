import { STATUS_CONTROL_CHANGE } from 'zds-pickers'
import { SHIFTER_CC_MESSAGE } from '../midi'
import { createReducer } from './utils'
import actionTypes from './actionTypes'

const defaultState = {
  controlId: 0,
  channel: 9,
  calibrationLow: 0,
  calibrationHigh: 127,
  curve: 0,
  polarity: 0,
  latching: 0,
  status: STATUS_CONTROL_CHANGE,
  value: SHIFTER_CC_MESSAGE,
  active: false,
  lit: false,
}

const modifyField = fieldName => (
  state,
  action, //
) => state.controlId === action.controlId //
  ? { ...state, [fieldName]: action[fieldName] }
  : { ...state }

export default createReducer(defaultState, {
  [actionTypes.CHANGE_INPUT_CONTROL_CALIBRATION_HIGH]: modifyField('calibrationHigh'),
  [actionTypes.CHANGE_INPUT_CONTROL_CALIBRATION_LOW]: modifyField('calibrationLow'),
  [actionTypes.CHANGE_INPUT_CONTROL_CHANNEL]: modifyField('channel'),
  [actionTypes.CHANGE_INPUT_CONTROL_LATCHING]: modifyField('latching'),
  [actionTypes.CHANGE_INPUT_CONTROL_POLARITY]: modifyField('polarity'),
  [actionTypes.CHANGE_INPUT_CONTROL_VALUE]: modifyField('value'),
})
