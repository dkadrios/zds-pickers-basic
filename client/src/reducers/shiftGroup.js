/* eslint-disable no-plusplus, no-mixed-operators */
import React from 'react'
import shallowEqual from 'shallowequal'
import { statuses } from 'zds-pickers'
import { omit } from '../core/fp/objects'
import { chunk } from '../core/fp/arrays'
import InputStatusIcon from '../components/InputStatusIcon'
import { SORT_ASC, SORT_BY_ALL, SORT_ON_INPUT } from '../core/consts'
import { createReducer } from './utils'
import actionTypes from './actionTypes'
import shiftEntry from './shiftEntry'

const defaultState = {
  active: false,
  editing: false,
  editQueue: {
    entryId: -1,
    input: {
      channel: 9,
      status: 9,
      value: 0,
    },
    output: {
      channel: 9,
      status: 9,
      value: 0,
    },
  },
  /**
   * Do not confuse selectedEntryId and selectedRows
   *
   * selectedEntryId is the item being edited
   * selectedRows is used for multiple selection and drag and drop
   */
  selectedEntryId: NaN,
  selectedRows: [],
  sortOn: SORT_ON_INPUT,
  sortBy: SORT_BY_ALL,
  sortDir: SORT_ASC,
}

const modifyField = fieldName => (state, action) => ({
  ...state,
  [fieldName]: action[fieldName],
})

const addEntry = (state, action) => ({
  ...state,
  editQueue: shiftEntry({}, { ...action, channel: state.channel, entryId: -1 }),
  editing: true,
})

const removeEntry = (state, { entryId: removeEntryId }) => {
  let newState = {
    ...state,
    selectedEntryId: NaN,
    selectedRows: [],
  }

  if (Array.isArray(removeEntryId)) {
    newState = {
      ...newState,
      entries: state.entries
        .filter(({ entryId }) => !removeEntryId.includes(entryId))
        // Realign ids to indexes
        .map((entry, idx) => ({ ...entry, entryId: idx })),
    }
  } else {
    newState = {
      ...newState,
      entries: state.entries
        .filter(({ entryId }) => entryId !== removeEntryId)
        // Realign ids to indexes
        .map((entry, idx) => ({ ...entry, entryId: idx })),
    }
  }

  return newState
}

const editEntry = (state, { editEntryIdx }) => ({
  ...state,
  editQueue: state.entries.filter(({ entryId }) => entryId === editEntryIdx)[0],
  editing: true,
})

const saveEdit = (state, { editQueue }) => {
  const localEditQueue = editQueue || state.editQueue
  const { entries } = state
  const { entryId } = localEditQueue
  return {
    ...state,
    editing: false,
    selectedEntryId: entryId === -1 ? entries.length : entryId,
    selectedRows: [entryId === -1 ? entries.length : entryId],
    entries:
      entryId === -1 // Adding new entry
        ? [...entries, { ...localEditQueue, entryId: entries.length }]
        : entries // Editing existing entry
          .map(entry => (entry.entryId === entryId ? { ...localEditQueue } : entry)),
  }
}

const cancelEdit = state => ({
  ...state,
  editing: false,
})

const passThrough = (state, action) => ({
  ...state,
  editQueue: shiftEntry(state.editQueue, action),
})

const receivedGroup = (state, { entries }) => ({
  ...defaultState,
  ...state,
  entries: chunk(4)(entries).map((
    entryData,
    entryId, //
  ) => shiftEntry({}, { type: actionTypes.RECEIVED_ENTRY, entryId, entryData })),
})

const changeGroupSort = (state, { sortOn, sortBy }) => {
  const newState = {
    ...state,
    sortOn,
    sortBy,
  }

  // selectedEntryId can be NaN, which is never equal to another NaN
  const ignoreForEquivalence = omit('selectedEntryId')

  if (shallowEqual(ignoreForEquivalence(state), ignoreForEquivalence(newState))) {
    // If there was no change then assume user clicked an already sorted col.
    // Toggle the sort direction.
    newState.sortDir = 1 - newState.sortDir
  } else {
    newState.sortDir = SORT_ASC
  }
  return newState
}

const selectShiftEntry = (state, { selectedEntryId }) => ({
  ...state,
  selectedEntryId,
  selectedRows: [selectedEntryId],
})

const metaSelectEntry = (state, { meta, shift, sortedIds, idx }) => {
  let selectedRows = [...state.selectedRows]
  const entryId = sortedIds[idx]
  let i

  if (meta) {
    // Add or remove the item from selected, but keep existing items
    if (selectedRows.includes(entryId)) {
      selectedRows = selectedRows.filter(id => id !== entryId)
    } else {
      selectedRows.push(entryId)
    }
  } else if (shift && selectedRows.length === 1) {
    // If only a single item is selected, select all the items in between
    const start = sortedIds.indexOf(selectedRows[0])
    const end = sortedIds.indexOf(entryId)
    let loopEntryId
    i = start
    while (i !== end) {
      if (i > end) {
        i -= 1
      } else {
        i += 1
      }
      loopEntryId = sortedIds[i]

      /* istanbul ignore else */
      if (selectedRows.indexOf(loopEntryId) === -1) {
        selectedRows.push(loopEntryId)
      }
    }
  }

  return { ...state, selectedRows, selectedEntryId: NaN }
}

const selectAll = state => ({
  ...state,
  selectedRows: state.entries.map(({ entryId }) => entryId),
})

const selectNone = state => ({
  ...state,
  selectedRows: [],
  selectedEntryId: NaN,
})

export const gridFriendlyData = ({ input, output }) => ({
  input: {
    ...input,
    statusLabel: statuses.reduce((acc, { label, value }) => (value === input.status ? label : acc), ''),
    channel: input.channel,
  },
  output: {
    ...output,
    statusLabel: statuses.reduce((acc, { label, value }) => (value === output.status ? label : acc), ''),
    channel: output.channel,
  },
  indicator: <InputStatusIcon {...input} />,
})

export default createReducer(defaultState, {
  [actionTypes.CHANGE_GROUP_CHANNEL]: modifyField('channel'),
  [actionTypes.CHANGE_GROUP_VALUE]: modifyField('value'),
  [actionTypes.ADD_ENTRY]: addEntry,
  [actionTypes.REMOVE_ENTRY]: removeEntry,
  [actionTypes.EDIT_ENTRY]: editEntry,
  [actionTypes.SAVE_ENTRY_EDIT]: saveEdit,
  [actionTypes.CANCEL_ENTRY_EDIT]: cancelEdit,
  [actionTypes.CHANGE_ENTRY_STATUS]: passThrough,
  [actionTypes.CHANGE_ENTRY_CHANNEL]: passThrough,
  [actionTypes.CHANGE_ENTRY_VALUE]: passThrough,
  [actionTypes.RECEIVED_GROUP]: receivedGroup,
  [actionTypes.SORT_GROUP]: changeGroupSort,
  [actionTypes.SELECT_SHIFT_ENTRY]: selectShiftEntry,
  [actionTypes.SELECT_ALL_ENTRIES]: selectAll,
  [actionTypes.CLEAR_ALL_SELECTIONS]: selectNone,
  [actionTypes.ENTRY_META_CLICKED]: metaSelectEntry,
})
