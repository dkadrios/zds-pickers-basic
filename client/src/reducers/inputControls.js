/* eslint-disable no-plusplus, no-bitwise */
import chunk from 'lodash/chunk'
import { MASK_CHANNEL, MASK_STATUS, assertRange } from 'zds-pickers'
import { createReducer } from '../utils'
import { MASK_CURVE, MASK_LATCHING, MASK_POLARITY } from '../midi'
import inputControl from './inputControl'
import actionTypes from './actionTypes'

export const actions = {
  askForControls: () => ({
    type: actionTypes.GET_SYSEX_CONTROLS,
  }),

  receivedControls: controlData => ({
    type: actionTypes.RECEIVED_CONTROLS,
    controlData,
  }),

  changeControlType: () => ({
    type: actionTypes.CHANGE_CONTROL_TYPE,
  }),

  changeInputControlChannel: (controlId, channel) => ({
    type: actionTypes.CHANGE_INPUT_CONTROL_CHANNEL,
    controlId,
    channel: assertRange(channel, 15, 0),
  }),

  changeInputControlLatching: (controlId, latching) => ({
    type: actionTypes.CHANGE_INPUT_CONTROL_LATCHING,
    controlId,
    latching,
  }),

  changeInputControlPolarity: (controlId, polarity) => ({
    type: actionTypes.CHANGE_INPUT_CONTROL_POLARITY,
    controlId,
    polarity,
  }),

  changeInputControlValue: (controlId, value) => ({
    type: actionTypes.CHANGE_INPUT_CONTROL_VALUE,
    controlId,
    value: assertRange(value, 127, 0),
  }),

  receivedInternalState: packet => ({
    type: actionTypes.RECEIVED_STATE,
    packet,
  }),
}

const modifyControl = (
  state,
  action, //
) => state.map(control => inputControl(control, action))

const receivedControls = (
  state,
  { controlData }, //
) => chunk(controlData, 5) //
  .map(([status, value, flags, calibrationLow, calibrationHigh], i) => inputControl(
    {
      controlId: i,
      channel: status & MASK_CHANNEL,
      calibrationLow,
      calibrationHigh,
      curve: (flags & MASK_CURVE) >> 2,
      polarity: (flags & MASK_POLARITY) >> 1,
      latching: flags & MASK_LATCHING,
      status: ((status & MASK_STATUS) >> 4) | 8,
      value,
      active: false,
      lit: false,
    },
    { type: 'NO_ACTION' },
  ))

const receivedInternalState = (state, { packet }) => {
  const [numControls, ...rest] = packet
  rest.pop() // Settings flag
  const [active, lit] = chunk(rest.map(val => Boolean(val)), numControls)

  return state.map((control, idx) => ({ ...control, active: active[idx], lit: lit[idx] }))
}

const handleShifterUnplugged = () => []

export default createReducer([], {
  [actionTypes.CHANGE_INPUT_CONTROL_CHANNEL]: modifyControl,
  [actionTypes.CHANGE_INPUT_CONTROL_LATCHING]: modifyControl,
  [actionTypes.CHANGE_INPUT_CONTROL_POLARITY]: modifyControl,
  [actionTypes.CHANGE_INPUT_CONTROL_VALUE]: modifyControl,
  [actionTypes.RECEIVED_CONTROLS]: receivedControls,
  [actionTypes.SHIFTER_MISSING]: handleShifterUnplugged,
  [actionTypes.FACTORY_RESET]: handleShifterUnplugged,
  [actionTypes.RECEIVED_STATE]: receivedInternalState,
})
