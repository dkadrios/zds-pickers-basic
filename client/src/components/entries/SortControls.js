/* eslint-disable jsx-a11y/anchor-is-valid */

import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames/bind'
import { SORT_ASC, SORT_ON_INPUT, SORT_ON_OUTPUT } from '../../reducers/shiftGroup'
import * as styles from '../../styles/shiftGroupTable.scss'
import { sortShape } from '../../shapes'

const cx = classNames.bind(styles)

const SortControls = ({ groupId, isInput, field, sortOn, sortBy, sortDir, children, changeSort }) => {
  const handleClick = (event) => {
    event.preventDefault()
    changeSort(groupId, isInput ? SORT_ON_INPUT : SORT_ON_OUTPUT, field)
  }

  const icon = () => (sortDir === SORT_ASC ? 'keyboard_arrow_down' : 'keyboard_arrow_up')

  const current = () => field === sortBy //
    && ((isInput && sortOn === SORT_ON_INPUT) || (!isInput && sortOn === SORT_ON_OUTPUT))

  return (
    <a
      className={cx({ sortControl: true, active: current() })}
      href="#"
      onClick={handleClick}
    >
      {children}
      <i className="material-icons">{icon()}</i>
    </a>
  )
}

SortControls.propTypes = {
  groupId: PropTypes.number.isRequired,
  isInput: PropTypes.bool,
  field: sortShape.sortBy, // eslint-disable-line
  ...sortShape,
  changeSort: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
}

SortControls.defaultProps = {
  isInput: false,
}

export default SortControls
