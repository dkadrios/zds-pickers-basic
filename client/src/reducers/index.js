import { combineReducers } from 'redux'
import { reducer as midi } from 'redux-midi-fork'
import { reducer as form } from 'redux-form'
import hardwareTest from './hardwareTest'
import inputControls from './inputControls'
import mappings from './mappings'
import shifter from './shifter'
import shiftGroups from './shiftGroups'
import user from './user'
import version from './version'

const reducers = {
  form,
  hardwareTest,
  inputControls,
  mappings,
  midi,
  shifter,
  shiftGroups,
  user,
  version,
}

export default combineReducers(reducers)
