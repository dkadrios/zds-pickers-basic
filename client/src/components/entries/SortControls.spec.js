import React from 'react'
import flow from 'lodash/fp/flow'
import set from 'lodash/fp/set'
import { SORT_ASC, SORT_BY_CHANNEL, SORT_DESC, SORT_ON_INPUT, SORT_ON_OUTPUT } from '../../reducers/shiftGroup'
import SortControls from './SortControls'

const changeSort = jest.fn()

const groupId = 0
const field = SORT_BY_CHANNEL

const props = {
  groupId,
  isInput: true,
  field,
  changeSort,
  children: <div>content</div>,
  sortOn: SORT_ON_INPUT,
  sortBy: SORT_BY_CHANNEL,
  sortDir: SORT_ASC,
}

describe('SortControls tests', () => {
  beforeEach(jest.clearAllMocks)

  it('renders correctly', () => {
    shallowExpect(SortControls)(props).toMatchSnapshot()
  })

  it('renders correctly when sorting desc', () => {
    shallowExpect(SortControls)(set('sortDir', SORT_DESC)(props)).toMatchSnapshot()
  })

  it('should also sort on output', () => {
    shallowExpect(SortControls)(flow(
      set('isInput', false),
      set('sortOn', SORT_ON_OUTPUT),
    )(props)).toMatchSnapshot()
  })

  it('changes sort on click', () => {
    const wrapper = shallow(<SortControls {...props} />)
    const preventDefault = jest.fn()
    const event = { preventDefault }
    wrapper.simulate('click', event)
    expect(preventDefault).toHaveBeenCalled()
    expect(changeSort).toHaveBeenCalledWith(groupId, SORT_ON_INPUT, field)
  })

  it('changes output sort on click', () => {
    const wrapper = shallow(<SortControls {...set('isInput', false)(props)} />)
    const preventDefault = jest.fn()
    const event = { preventDefault }
    wrapper.simulate('click', event)
    expect(preventDefault).toHaveBeenCalled()
    expect(changeSort).toHaveBeenCalledWith(groupId, SORT_ON_OUTPUT, field)
  })
})
