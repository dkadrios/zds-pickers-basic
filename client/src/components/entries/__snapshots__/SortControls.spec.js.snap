// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortControls tests renders correctly 1`] = `
<a
  className="sortControl active"
  href="#"
  onClick={[Function]}
>
  <div>
    content
  </div>
  <i
    className="material-icons"
  >
    keyboard_arrow_down
  </i>
</a>
`;

exports[`SortControls tests renders correctly when sorting desc 1`] = `
<a
  className="sortControl active"
  href="#"
  onClick={[Function]}
>
  <div>
    content
  </div>
  <i
    className="material-icons"
  >
    keyboard_arrow_up
  </i>
</a>
`;

exports[`SortControls tests should also sort on output 1`] = `
<a
  className="sortControl active"
  href="#"
  onClick={[Function]}
>
  <div>
    content
  </div>
  <i
    className="material-icons"
  >
    keyboard_arrow_down
  </i>
</a>
`;
